src/distributions/weighted/mod.rs

```diff
@@ -53,7 +53,7 @@ use core::fmt;
 ///
 /// ```
 /// use rand::prelude::*;
-/// use rand::distributions::WeightedIndex;
+/// use rand::distributions::weighted::WeightedIndex;
```

Comment:
If we remove this, I think we might as well remove the `Weighted` prefix and use `weighted::Index` and `weighted::Error` in the code.

Reply:
The full path is ridiculously long. But is `weighted::Index` not too confusing? @zroug

Reply:
I think it's fine. It is similar to `std::io::Error` and *ring*'s naming scheme.

Reply:
Usually I tend to use the shorter variant but `Index` is on the border to being to meaningless by itself. (It never makes sense to `use weighted::Index;` but it might make sense to `use std::io::Error;` if it is the only error type in the context.) On the other hand `WeightedIndex` makes it less convenient to just `use weighted;`)

If I would have to make a choice I would probably go with `WeightedIndex` because `Error` sounds like less specific term for `IOError` but `Index` doesn't sound like a less specific term for `WeightedIndex`.

Reply:
I think we could suggest to always use `use weighted;` and `weighted::Index`.

Reply:
I agree with @zroug; `Index` by itself is not meaningful. `Error` is sufficient but not much of an advantage by itself. I don't really have much to add that [wasn't already covered](https://github.com/rust-random/rand/pull/692#discussion_r260797930), except that for the most part we name things by what they do, not the method used (there are some exceptions under `rand::seq::index`, but these are internal only).

It wouldn't be impossible to combine the `WeightedIndex` types into a single struct with two different constructors, though it would require all bounds on the weight type regardless of which method was used (as well as a little extra memory usage for the binary search method).

There is no reason we need `Index` in the name; essentially this is a non-uniform discrete (integer) distribution.

Reply:
I removed the associated change (deleting the re-export) from this PR. This still needs addressing, but lets get this PR merged!